qubes/vm/dispvm.py

```diff
@@ -187,11 +189,131 @@ def __init__(self, app, xml, *args, **kwargs):
+        appvm.features["preload-dispvm"] = " ".join(preload_dispvm or [])
+        self.features["internal"] = False
+        await appvm.fire_event_async(
+            "domain-preloaded-dispvm-used", dispvm=self
+        )
+
     @qubes.events.handler("domain-load")
     def on_domain_loaded(self, event):
         """When domain is loaded assert that this vm has a template."""  # pylint: disable=unused-argument
         assert self.template
 
+    @qubes.events.handler("domain-start")
+    async def on_domain_started_dispvm(
+        self,
+        event,
+        **kwargs,
+    ):  # pylint: disable=unused-argument
+        """Pause preloaded domains as soon as basic services have started."""
+        if not self.is_preloaded():
+            return
+        # TODO:
+        # Marek: Test if pause isn't too early. Some services (especially:
+        #   gui-agent) may still be starting.  qubes.WaitForSession service may
+        #   help (ensure to use async handler to not block qubesd while waiting
+        #   on it).
+        # TODO:
+        # Ben:
+        #   Test if pause isn't too late, what if application autostarts, will
+        #   it open before the qube is paused?
+        # Marek:
+        #   Yes, it will. Theoretically there is an "invisible" mode for
+        #   gui-daemon for situation like this (it was used for very old
+        #   implementation of DispVM that also kinda preloaded it). But there
+        #   is no support for flipping it in runtime, gui-daemon needs to be
+        #   restarted for that, so that's a broader change to use it in this
+        #   version. Maybe later, I'd say it's okay to ignore this issue for
+        #   now.
+        # Ben:
+        #   I set xterm.dekstop to autostart, tested that it autostarted first
+        #   and then tested if pause isn't too late:
+        #     import qubesadmin
+        #     domains = qubesadmin.Qubes().domains
+        #     q = domains['q']
+        #     if q.run_service_for_stdio("qubes.WaitForSession"):
+        #         q.pause()
+        #   XTerm did not appear, and this is on a minimal qube that has no
+        #   heavy service that delays the start. As soon as I did
+        #   'q.unpause()', the application window appeared.
+        no_gui_sleep = 15
+        gui_timeout = 30
+        gui = self.features.get("gui", None)
+        if not gui:
+            await asyncio.sleep(no_gui_sleep)
+            await self.pause()
+            return
+        proc = None
+        try:
+            proc = await asyncio.wait_for(
+                self.run_service_for_stdio(
+                    "qubes.WaitForSession",
+                    user=self.default_user,
+                    stdout=subprocess.DEVNULL,
+                    stderr=subprocess.DEVNULL,
+                ),
+                timeout=gui_timeout,
+            )
+        except asyncio.TimeoutError:
+            ## TODO: should timeout be treated as an error/qubes.exc?
```

Comment:
Throwing an exception from this event handler, will cause VM to be killed. I'm not sure if that's a good idea, especially with hardcoded timeouts. If not killing, it will pause a VM that potentially is still in starting, right? If that's harmless then just log a message but don't kill. But if that causes some issues (for example, systemd on unpause considers various service start timeouts already expired), then probably better kill and maybe make timeout adjustable (or derive from qrexec_timeout property?)